File: src/platforms/evdev-rs/platform.cpp
/*
 * Copyright © Canonical Ltd.
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License version 2 or 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "platform.h"
#include "platform_bridge.h"
#include "input_report.h"

#include <mir/optional_value.h>
#include <mir/dispatch/dispatchable.h>
#include <mir/dispatch/multiplexing_dispatchable.h>
#include <mir/dispatch/readable_fd.h>
#include <mir_platforms_evdev_rs/src/lib.rs.h>
#include <rust/cxx.h>
#include <mir/input/pointer_settings.h>
#include <mir/input/touchpad_settings.h>
#include <mir/input/touchscreen_settings.h>
#include <mir/input/input_device_info.h>
#include <mir/dispatch/action_queue.h>
#include <mir/log.h>

#include <unistd.h>

namespace mi = mir::input;
namespace miers = mi::evdev_rs;
namespace md = mir::dispatch;

/// Thin observer that forwards device lifecycle events to the Rust platform
/// via an ActionQueue (ensuring they run on the input dispatch thread).
class InputDeviceObserver : public mir::Device::Observer
{
public:
    InputDeviceObserver(
        std::string devnode,
        uint64_t devnum,
        std::shared_ptr<md::ActionQueue> device_queue,
        std::weak_ptr<miers::Platform::Self> platform_self)
        : devnode{std::move(devnode)},
          devnum{devnum},
          device_queue{std::move(device_queue)},
          platform_self{std::move(platform_self)}
    {
    }

    void activated(mir::Fd&& device_fd) override;
    void suspended() override;
    void removed() override;

private:
    std::string devnode;
    uint64_t devnum;
    std::shared_ptr<md::ActionQueue> device_queue;
    std::weak_ptr<miers::Platform::Self> platform_self;
};

namespace
{

template <typename Impl>
class InputDevice : public mi::InputDevice
{
public:
    explicit InputDevice(rust::Box<Impl>&& impl) : impl(std::move(impl)) {}

    void start(mi::InputSink* sink, mi::EventBuilder* event_builder) override
    {
        impl->start(sink, event_builder);
    }

    void stop() override
    {
        impl->stop();
    }

    mi::InputDeviceInfo get_device_info() override
    {
        auto device_info = impl->get_device_info();
        if (!device_info->valid())
            mir::log_error("Unable to get device info for the input device.");

        return {
            std::string(device_info->name()),
            std::string(device_info->unique_id()),
            mi::DeviceCapabilities(device_info->capabilities()),
        };
    }

    mir::optional_value<mi::PointerSettings> get_pointer_settings() const override
    {
        auto pointer_settings = impl->get_pointer_settings();
        if (!pointer_settings->is_set)
            return mir::optional_value<mi::PointerSettings>{};

        if (pointer_settings->has_error)
        {
            mir::log_warning("Unable to get pointer settings for the input device.");
            return mir::optional_value<mi::PointerSettings>{};
        }

        mi::PointerSettings result;
        result.handedness = static_cast<MirPointerHandedness>(pointer_settings->handedness);
        result.acceleration = static_cast<MirPointerAcceleration>(pointer_settings->acceleration);
        result.cursor_acceleration_bias = pointer_settings->cursor_acceleration_bias;
        result.horizontal_scroll_scale = pointer_settings->horizontal_scroll_scale;
        result.vertical_scroll_scale = pointer_settings->vertical_scroll_scale;
        return result;
    }

    void apply_settings(mi::PointerSettings const& pointer_settings) override
    {
        miers::SetPointerSettingsData pointer_settings_c;
        pointer_settings_c.handedness = static_cast<MirPointerHandedness>(pointer_settings.handedness);
        pointer_settings_c.acceleration = static_cast<MirPointerAcceleration>(pointer_settings.acceleration);
        pointer_settings_c.cursor_acceleration_bias = pointer_settings.cursor_acceleration_bias;
        pointer_settings_c.horizontal_scroll_scale = pointer_settings.horizontal_scroll_scale;
        pointer_settings_c.vertical_scroll_scale = pointer_settings.vertical_scroll_scale;
        impl->set_pointer_settings(pointer_settings_c);
    }

    mir::optional_value<mi::TouchpadSettings> get_touchpad_settings() const override
    {
        // TODO(mattkae): Handle touch gestures and settings
        return mir::optional_value<mi::TouchpadSettings>{};
    }

    void apply_settings(mi::TouchpadSettings const&) override
    {
        // TODO(mattkae): Handle touch gestures and settings
    }

    mir::optional_value<mi::TouchscreenSettings> get_touchscreen_settings() const override
    {
        // TODO(mattkae): Handle touch gestures and settings
        return mir::optional_value<mi::TouchscreenSettings>{};
    }

    void apply_settings(mi::TouchscreenSettings const&) override
    {
        // TODO(mattkae): Handle touch gestures and settings
    }

private:
    rust::Box<Impl> impl;
};

} // namespace

class miers::Platform::Self
{
public:
    Self(Platform* platform,
         std::shared_ptr<ConsoleServices> const& console,
         std::shared_ptr<InputDeviceRegistry> const& input_device_registry,
         std::shared_ptr<mi::InputReport> const& report)
        : bridge{std::make_shared<PlatformBridge>(platform, console)},
          platform_impl{evdev_rs_create(
              bridge,
              input_device_registry,
              std::make_unique<miers::InputReport>(report))},
          dispatchable{std::make_shared<md::MultiplexingDispatchable>()}
    {
    }

    std::shared_ptr<PlatformBridge> bridge;
    rust::Box<miers::PlatformRs> platform_impl;
    std::shared_ptr<md::ReadableFd> libinput_dispatch;
    std::shared_ptr<md::ReadableFd> udev_dispatch;
    std::shared_ptr<md::MultiplexingDispatchable> dispatchable;
    std::shared_ptr<md::ActionQueue> device_queue;
};

void InputDeviceObserver::activated(mir::Fd&& device_fd)
{
    mir::log_info("evdev-rs: observer activated() for %s (devnum=%lu, fd=%d)",
                   devnode.c_str(), static_cast<unsigned long>(devnum), static_cast<int>(device_fd));

    auto const locked = platform_self.lock();
    if (!locked)
    {
        mir::log_error("evdev-rs: activated() ignored — platform is not allocated");
        return;
    }

    auto& platform_impl = locked->platform_impl;
    if (!platform_impl->is_running())
    {
        mir::log_info("evdev-rs: activated() ignored — platform not running");
        return;
    }

    mir::log_info("evdev-rs: enqueuing on_device_activated for %s (duped fd=%d)", devnode.c_str(), duped);
    device_queue->enqueue(
        [devnode = devnode, devnum = devnum, device_fd = device_fd, &platform_impl = platform_impl]()
        {
            // dup() so that mir::Fd's destructor closes the original without
            // invalidating the fd we hand to libinput via open_restricted().
            int const duped = ::dup(static_cast<int>(device_fd));
            if (duped < 0)
            {
                mir::log_error("evdev-rs: dup() failed in activated() for %s", devnode.c_str());
                return;
            }

            if (!platform_impl->is_running())
            {
                mir::log_info("evdev-rs: on_device_activated dequeued but platform not running, closing fd=%d", fd);
                ::close(duped);
                return;
            }
            platform_impl->on_device_activated(devnode, devnum, duped);
        });
}

void InputDeviceObserver::suspended()
{
    mir::log_info("evdev-rs: observer suspended() for %s (devnum=%lu)",
                   devnode.c_str(), static_cast<unsigned long>(devnum));

    auto const locked = platform_self.lock();
    if (!locked)
    {
        mir::log_error("evdev-rs: suspended() ignored — platform is not allocated");
        return;
    }

    auto& platform_impl = locked->platform_impl;

    if (!platform_impl->is_running())
    {
        mir::log_info("evdev-rs: suspended() ignored — platform not running");
        return;
    }

    device_queue->enqueue(
        [devnode = devnode, devnum = devnum, &platform_impl = platform_impl]()
        {
            if (!platform_impl->is_running())
                return;
            platform_impl->on_device_suspended(devnode, devnum);
        });
}

void InputDeviceObserver::removed()
{
    mir::log_info("evdev-rs: observer removed() for %s (devnum=%lu)",
                   devnode.c_str(), static_cast<unsigned long>(devnum));

    auto const locked = platform_self.lock();
    if (!locked)
    {
        mir::log_error("evdev-rs: removed() ignored — platform is not allocated");
        return;
    }

    auto& platform_impl = locked->platform_impl;

    if (!platform_impl->is_running())
    {
        mir::log_info("evdev-rs: removed() ignored — platform not running");
        return;
    }

    device_queue->enqueue(
        [devnode = devnode, devnum = devnum, &platform_impl = platform_impl]()
        {
            if (!platform_impl->is_running())
                return;
            platform_impl->on_device_removed(devnode, devnum);
        });
}

miers::Platform::Platform(
    std::shared_ptr<ConsoleServices> const& console,
    std::shared_ptr<InputDeviceRegistry> const& input_device_registry,
    std::shared_ptr<mi::InputReport> const& report)
    : self(std::make_shared<Self>(this, console, input_device_registry, report))
{
}

std::shared_ptr<mir::dispatch::Dispatchable> miers::Platform::dispatchable()
{
    return self->dispatchable;
}

void miers::Platform::start()
{
    // Clean up any stale watches from a previous start()/stop() cycle.
    if (self->device_queue)
    {
        self->dispatchable->remove_watch(self->device_queue);
        self->device_queue.reset();
    }
    if (self->udev_dispatch)
    {
        self->dispatchable->remove_watch(self->udev_dispatch);
        self->udev_dispatch.reset();
    }
    if (self->libinput_dispatch)
    {
        self->dispatchable->remove_watch(self->libinput_dispatch);
        self->libinput_dispatch.reset();
    }

    // Create the device_queue before start() because Rust's start()
    // enumerates existing udev devices, which triggers acquire_device()
    // and may fire activated() synchronously — needing the queue.
    self->device_queue = std::make_shared<md::ActionQueue>();

    if (!self->platform_impl->start())
    {
        self->device_queue.reset();
        return;
    }

    auto const libinput_fd = self->platform_impl->libinput_fd();
    if (libinput_fd < 0)
    {
        mir::log_error("evdev-rs platform: failed to obtain libinput fd after start(); resetting platform state");
        self->platform_impl->stop();
        self->device_queue.reset();
        return;
    }

    self->libinput_dispatch = std::make_shared<md::ReadableFd>(
        mir::Fd{mir::IntOwnedFd{libinput_fd}},
        [this]() { self->platform_impl->process(); });

    auto const udev_fd = self->platform_impl->udev_monitor_fd();
    if (udev_fd < 0)
    {
        mir::log_error("evdev-rs platform: failed to obtain udev monitor fd; resetting platform state");
        self->platform_impl->stop();
        self->device_queue.reset();
        return;
    }

    self->udev_dispatch = std::make_shared<md::ReadableFd>(
        mir::Fd{mir::IntOwnedFd{udev_fd}},
        [this]() { self->platform_impl->process_udev_events(); });

    self->dispatchable->add_watch(self->libinput_dispatch);
    self->dispatchable->add_watch(self->device_queue);
    self->dispatchable->add_watch(self->udev_dispatch);
}

void miers::Platform::continue_after_config()
{
    self->platform_impl->continue_after_config();
}

void miers::Platform::pause_for_config()
{
    self->platform_impl->pause_for_config();
}

void miers::Platform::stop()
{
    // Remove watches before stopping Rust, because stop() closes the libinput
    // and udev fds. Calling remove_watch() after the fds are closed causes
    // epoll_ctl(EPOLL_CTL_DEL) to return EBADF and throw.
    if (self->udev_dispatch)
    {
        self->dispatchable->remove_watch(self->udev_dispatch);
        self->udev_dispatch.reset();
    }
    if (self->device_queue)
    {
        self->dispatchable->remove_watch(self->device_queue);
        self->device_queue.reset();
    }
    if (self->libinput_dispatch)
    {
        self->dispatchable->remove_watch(self->libinput_dispatch);
        self->libinput_dispatch.reset();
    }

    self->platform_impl->stop();
}

std::unique_ptr<mir::Device::Observer> miers::Platform::create_device_observer(
    std::string const& devnode, uint64_t devnum)
{
    return std::make_unique<InputDeviceObserver>(
        devnode, devnum, self->device_queue, self);
}

std::shared_ptr<mi::InputDevice> miers::Platform::create_input_device(int device_id) const
{
    return std::make_shared<::InputDevice<LibinputDevice>>(self->platform_impl->create_input_device(device_id));
}

File: src/platforms/evdev-rs/src/platform.rs
/*
 * Copyright © Canonical Ltd.
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License version 2 or 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use crate::device::{LibinputDevice, LibinputDeviceState, ScrollState};
use crate::event_processing::process_libinput_events;
use crate::fd_store::FdStore;
use crate::udev_monitor::{UdevEventType, UdevMonitor};
use std::collections::HashSet;
use std::os::fd::{AsRawFd, FromRawFd, OwnedFd};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::sync::Mutex;

pub struct PlatformRs {
    /// The platform bridge provides access to functionality that must happen in C++.
    bridge: cxx::SharedPtr<crate::PlatformBridge>,

    /// The input device registry is used for registering and unregistering input devices.
    device_registry: cxx::SharedPtr<crate::InputDeviceRegistry>,

    report: cxx::UniquePtr<crate::InputReport>,

    state: Option<Arc<Mutex<LibinputDeviceState>>>,

    /// Whether the platform is currently running. Owned by Rust, queried by C++.
    running: Arc<AtomicBool>,

    /// Set of devnums we know about (pending or active), to deduplicate udev events.
    known_devnums: HashSet<u64>,

    /// Rust-owned udev monitor for input device hotplug events.
    udev_monitor: Option<UdevMonitor>,

    /// Manages pre-acquired and backup file descriptors for device nodes.
    fd_store: Arc<Mutex<FdStore>>,
}

impl PlatformRs {
    pub fn new(
        bridge: cxx::SharedPtr<crate::PlatformBridge>,
        device_registry: cxx::SharedPtr<crate::InputDeviceRegistry>,
        report: cxx::UniquePtr<crate::InputReport>,
    ) -> Self {
        PlatformRs {
            bridge,
            device_registry,
            report,
            state: None,
            running: Arc::new(AtomicBool::new(false)),
            known_devnums: HashSet::new(),
            udev_monitor: None,
            fd_store: Arc::new(Mutex::new(FdStore::new())),
        }
    }

    /// Start the input platform.
    ///
    /// Creates the libinput context using the path-based API so that device
    /// acquisition (via `path_add_device`) is driven by the `UdevMonitor`.
    /// This avoids `udev_assign_seat()`, which would call `open_restricted()`
    /// synchronously while the GLib main loop still needs to process the logind
    /// `TakeDevice` D-Bus replies.
    ///
    /// Returns `true` if a fresh context was created, `false` if the platform
    /// was already started.
    pub fn start(&mut self) -> bool {
        if self.state.is_some() {
            return false;
        }

        println!("Starting the evdev-rs platform");

        let libinput =
            input::Libinput::new_from_path(crate::libinput_interface::LibinputInterfaceImpl {
                fd_store: self.fd_store.clone(),
            });

        self.state = Some(Arc::new(Mutex::new(LibinputDeviceState {
            libinput,
            known_devices: Vec::new(),
            next_device_id: 0,
            scroll_state: ScrollState {
                x_accum: 0.0,
                y_accum: 0.0,
                x_scroll_scale: 1.0,
                y_scroll_scale: 1.0,
            },
        })));

        self.running.store(true, Ordering::Release);
        self.known_devnums.clear();

        // Create the udev monitor. The closure handles initial device enumeration.
        let mut initial_devices: Vec<(String, u64, String)> = Vec::new();
        self.udev_monitor = UdevMonitor::new(|_event_type, devnode, devnum, sysname| {
            initial_devices.push((devnode.to_string(), devnum, sysname.to_string()));
        });

        // Process the initially-enumerated devices through on_udev_event.
        for (devnode, devnum, sysname) in initial_devices {
            self.on_udev_event(UdevEventType::Added, &devnode, devnum, &sysname);
        }

        true
    }

    /// Add a device to the libinput context by path.
    ///
    /// Called internally after a device fd has been acquired and stored in
    /// the bridge's pending-fd map. libinput will call `open_restricted()`,
    /// which retrieves the pre-acquired fd non-blockingly via `claim_pending_fd`.
    pub fn path_add_device(&mut self, devnode: &str) {
        let Some(state_arc) = self.state.as_mut() else {
            println!("PlatformRs::path_add_device: platform not started");
            return;
        };
        match state_arc.lock() {
            Ok(mut state) => {
                println!("evdev-rs: libinput path_add_device({})", devnode);
                state.libinput.path_add_device(devnode);
            }
            Err(_) => {
                println!("PlatformRs::path_add_device: state mutex poisoned");
            }
        }
    }

    /// Remove a device from the libinput context by path.
    ///
    /// Called internally when a device is suspended (VT switch) or
    /// removed (hotplug disconnect).
    pub fn path_remove_device(&mut self, devnode: &str) {
        let Some(state_arc) = self.state.as_mut() else {
            return;
        };
        match state_arc.lock() {
            Ok(mut state) => {
                let index = state
                    .known_devices
                    .iter()
                    .position(|d| d.devnode == devnode);
                let Some(index) = index else {
                    println!(
                        "evdev-rs: path_remove_device({}) — not in known_devices",
                        devnode
                    );
                    return;
                };
                println!(
                    "evdev-rs: path_remove_device({}) — removing from libinput",
                    devnode
                );
                let device_info = state.known_devices.swap_remove(index);
                state.libinput.path_remove_device(device_info.device);

                let input_device = device_info.input_device.clone();
                let registry = self.device_registry.clone();
                std::thread::spawn(move || unsafe {
                    registry
                        .clone()
                        .pin_mut_unchecked()
                        .remove_device(&input_device);
                });
            }
            Err(_) => {
                println!("PlatformRs::path_remove_device: state mutex poisoned");
            }
        }
    }

    pub unsafe fn libinput_fd(&mut self) -> i32 {
        match self.state.as_mut() {
            Some(state_arc) => match state_arc.lock() {
                Ok(state) => state.libinput.as_raw_fd(),
                Err(_) => -1,
            },
            None => -1,
        }
    }

    pub fn continue_after_config(&self) {}
    pub fn pause_for_config(&self) {}

    /// Stop the input platform.
    ///
    /// This method will signal the input thread to stop and wait for it to finish.
    pub fn stop(&mut self) {
        self.running.store(false, Ordering::Release);
        self.known_devnums.clear();
        self.udev_monitor = None;

        // Note: When the main loop exits, we need to remove all devices from the device registry
        // or else Mir will try to free the devices later but we won't have access to the platform
        // code at that point. This results in a segfault.
        let Some(state_arc) = self.state.take() else {
            // Platform not started or already stopped.
            return;
        };

        // Collect the input devices to remove while holding the state lock.
        // We must NOT hold the state lock while calling remove_device(), because that
        // can cause an ABBA deadlock:
        //   - This thread: holds Rust state mutex, waiting for InputDeviceHub mutex (in remove_device)
        //   - A spawned add_device thread: holds InputDeviceHub mutex, waiting for Rust state mutex
        //     (in LibinputDevice::start)
        let devices_to_remove: Vec<cxx::SharedPtr<crate::InputDevice>> = {
            match state_arc.lock() {
                Ok(mut state_guard) => state_guard
                    .known_devices
                    .drain(..)
                    .map(|info| info.input_device)
                    .collect(),
                Err(poisoned) => {
                    // The mutex was poisoned by a previous panic, but we still drain
                    // known_devices from the inner state to avoid leaving them registered
                    // in the InputDeviceRegistry, which can lead to segfaults later.
                    let mut state_guard = poisoned.into_inner();
                    state_guard
                        .known_devices
                        .drain(..)
                        .map(|info| info.input_device)
                        .collect()
                }
            }
        };

        // Remove devices from the registry WITHOUT holding the Rust state mutex.
        for input_device in &devices_to_remove {
            // # Safety
            //
            // Because we need to use pin_mut_unchecked, this is unsafe.
            unsafe {
                self.device_registry
                    .clone()
                    .pin_mut_unchecked()
                    .remove_device(input_device);
            }
        }
    }

    pub fn create_input_device(&mut self, device_id: i32) -> Box<LibinputDevice> {
        let state_arc = match self.state.as_mut() {
            Some(s) => s.clone(),
            None => {
                eprintln!(
                    "PlatformRs::create_input_device: platform not started (device_id={})",
                    device_id
                );
                // Return a device that will fail gracefully when used.
                return Box::new(LibinputDevice {
                    device_id,
                    state: Arc::new(Mutex::new(LibinputDeviceState {
                        libinput: input::Libinput::new_from_path(
                            crate::libinput_interface::LibinputInterfaceImpl {
                                fd_store: self.fd_store.clone(),
                            },
                        ),
                        known_devices: Vec::new(),
                        next_device_id: 0,
                        scroll_state: ScrollState {
                            x_accum: 0.0,
                            y_accum: 0.0,
                            x_scroll_scale: 1.0,
                            y_scroll_scale: 1.0,
                        },
                    })),
                    bridge: self.bridge.clone(),
                });
            }
        };

        Box::new(LibinputDevice {
            device_id,
            state: state_arc,
            bridge: self.bridge.clone(),
        })
    }

    pub fn process(&mut self) {
        if self.state.is_none() {
            return;
        }

        if let Ok(mut state) = self.state.as_mut().unwrap().try_lock() {
            process_libinput_events(
                &mut state,
                self.device_registry.clone(),
                self.bridge.clone(),
                &self.report,
            );
        }
    }

    /// Returns whether the platform is currently running.
    /// Queried from C++ (InputDeviceObserver) and internally.
    pub fn is_running(&self) -> bool {
        self.running.load(Ordering::Acquire)
    }

    /// Returns the file descriptor for the udev monitor.
    /// C++ wraps this in a `ReadableFd` dispatchable.
    /// Returns -1 if the monitor is not available.
    pub fn udev_monitor_fd(&self) -> i32 {
        match &self.udev_monitor {
            Some(monitor) => monitor.fd(),
            None => -1,
        }
    }

    /// Process pending udev monitor events.
    /// Invoked from C++ when the udev monitor fd becomes readable.
    pub fn process_udev_events(&mut self) {
        if !self.running.load(Ordering::Acquire) {
            return;
        }

        // Take the monitor out temporarily to avoid borrow issues with self.
        let Some(mut monitor) = self.udev_monitor.take() else {
            return;
        };

        let mut events: Vec<(UdevEventType, String, u64, String)> = Vec::new();
        monitor.process_events(|event_type, devnode, devnum, sysname| {
            events.push((event_type, devnode.to_string(), devnum, sysname.to_string()));
        });

        // Put the monitor back before processing events (which need &mut self).
        self.udev_monitor = Some(monitor);

        for (event_type, devnode, devnum, sysname) in events {
            self.on_udev_event(event_type, &devnode, devnum, &sysname);
        }
    }

    /// Handle a udev event. Decides whether to acquire or release a device.
    ///
    /// Called internally by `process_udev_events()` (at runtime when the
    /// udev fd is readable) and by `start()` (during initial enumeration).
    pub fn on_udev_event(
        &mut self,
        event_type: UdevEventType,
        devnode: &str,
        devnum: u64,
        sysname: &str,
    ) {
        if !self.running.load(Ordering::Acquire) {
            return;
        }

        match event_type {
            UdevEventType::Added => {
                // Only process "event" devices (what libinput expects).
                if !sysname.starts_with("event") {
                    return;
                }

                // Deduplicate: skip if already pending or active.
                if self.known_devnums.contains(&devnum) {
                    println!(
                        "evdev-rs: on_udev_event Added devnode={} devnum={} — already known, skipping",
                        devnode, devnum
                    );
                    return;
                }

                println!(
                    "evdev-rs: on_udev_event Added devnode={} devnum={} — acquiring",
                    devnode, devnum
                );

                // Ask C++ to initiate device acquisition via ConsoleServices.
                if self.bridge.acquire_device(devnum, devnode) {
                    self.known_devnums.insert(devnum);
                }
            }
            UdevEventType::Removed => {
                println!(
                    "evdev-rs: on_udev_event Removed devnode={} devnum={}",
                    devnode, devnum
                );

                if !self.known_devnums.remove(&devnum) {
                    println!("evdev-rs: on_udev_event Removed — not in known_devnums, skipping");
                    return;
                }

                // Release the pending future if activated() never fired.
                self.bridge.release_pending_device(devnum);

                // If already active, remove from libinput and release.
                if self.bridge.has_device(devnum) {
                    if !devnode.is_empty() {
                        self.path_remove_device(devnode);
                    }
                    self.bridge.release_device(devnum);
                }

                if let Ok(mut store) = self.fd_store.lock() {
                    store.remove(devnode);
                }
            }
        }
    }

    /// Handle device activation (fd acquired from ConsoleServices).
    ///
    /// Invoked from C++ (on the input dispatch thread via ActionQueue)
    /// after `Device::Observer::activated()` has fired and the fd has
    /// been dup'd.
    pub fn on_device_activated(&mut self, devnode: &str, devnum: u64, fd: i32) {
        println!(
            "evdev-rs: on_device_activated devnode={} devnum={} fd={}",
            devnode, devnum, fd
        );

        if !self.running.load(Ordering::Acquire) {
            println!("evdev-rs: on_device_activated ignored — platform not running");
            return;
        }

        // Store the fd so that open_restricted() can claim it.
        match self.fd_store.lock() {
            Ok(mut store) => store.store(devnode, unsafe { OwnedFd::from_raw_fd(fd) }),
            Err(_) => {
                println!("evdev-rs: on_device_activated: fd_store mutex poisoned");
                return;
            }
        }

        // Transfer the Device from pending to active in C++.
        self.bridge.activate_pending_device(devnum);

        // Tell libinput about the new device.
        println!(
            "evdev-rs: on_device_activated calling path_add_device({})",
            devnode
        );
        self.path_add_device(devnode);
    }

    /// Handle device suspension (e.g. VT switch or lid close).
    ///
    /// Invoked from C++ (on the input dispatch thread via ActionQueue).
    pub fn on_device_suspended(&mut self, devnode: &str, devnum: u64) {
        println!(
            "evdev-rs: on_device_suspended devnode={} devnum={}",
            devnode, devnum
        );

        if !self.running.load(Ordering::Acquire) {
            println!("evdev-rs: on_device_suspended ignored — platform not running");
            return;
        }

        println!(
            "evdev-rs: on_device_suspended calling path_remove_device({})",
            devnode
        );

        self.bridge.release_device(devnum);
        self.path_remove_device(devnode);
    }

    /// Handle device removal (hotplug disconnect via observer).
    ///
    /// Invoked from C++ (on the input dispatch thread via ActionQueue).
    pub fn on_device_removed(&mut self, devnode: &str, devnum: u64) {
        println!(
            "evdev-rs: on_device_removed devnode={} devnum={}",
            devnode, devnum
        );

        if !self.running.load(Ordering::Acquire) {
            println!("evdev-rs: on_device_removed ignored — platform not running");
            return;
        }

        self.known_devnums.remove(&devnum);
        self.bridge.release_device(devnum);
        if let Ok(mut store) = self.fd_store.lock() {
            store.remove(devnode);
        }
        self.path_remove_device(devnode);
    }
}
